doc/design/ops/dist_train.md

```diff
@@ -0,0 +1,82 @@
+# Design Doc: Operation Graph Based Parameter Server
+
+## Abstract
+
+We propose an approach to implement the parameter server. In this
+approach, there is no fundamental difference between the trainer and
+the parameter server: they both run sub-graphs, but sub-graphs of
+different purposes.
+
+## Background
+
+The previous implementations of the parameter server does not run a
+sub-graph. parameter initialization, optimizer computation, network
+communication and checkpointing are implemented twice on both the
+trainer and the parameter server.
+
+It would be great if we can write code once and use them on both the
+trainer and the parameter server: reduces code duplication and
+improves extensibility. Given that after the current refactor, we are
+representing everything as a computing graph on the
+trainer. Representing everything as a computing graph on the parameter
+server becomes a natural extension.
+
+## Design
+
+### Graph Converter
+
+The *graph converter* converts the user-defined operation (OP) graph
+into sub-graphs to be scheduled on different nodes.
+
+1. The user-defined OP graph will be cut into sub-graphs of
+different purposes (e.g., trainer, parameter server) to run on
+different workers.
+
+1. OPs will be added to the subgraphs, so the subgraphs can
+communicate with each other. We will need these OPs: *send*, *recv*,
+*gradient accumulator*, *string accumulator*, *loop forever*.
+
+Below is an example of converting the user defined graph to the
+sub-graphs for the trainer and the parameter server:
+
+<img src="src/local-graph.png" width="300"/>
+
+After converting:
+
+<img src="src/dist-graph.png" width="500"/>
+
+1. The parameter variable W and it's optimizer subgraph are placed on the parameter server.
+1. Operators are added to the sub-graphs.
+   - *send* operator sends data and sender's address to the destination.
```

Comment:
Question: if there are multiple parameters(or variables) to send to parameter server, are we:

1. create multiple `Send` operators for each variable or,
2. create one `Hash` operator to divide parameters equally and one `Sender` operator to do send.

Reply:
And also, maybe add some description about the `send`, `recv` operators like:

- Send:
    - Inputs:
    - Outputs:
    - Description:

Reply:
The same confusion with @typhoonzero , maybe we need an operator to sharded parameters.

For others, from @typhoonzero 

> create multiple Send operators for each variable

Maybe we only need one `Sender` operator. If we have too many parameters, too much `Sender` operator will cased too much connection to parameter server.

Reply:
@typhoonzero @Yancey1989 sorry, the PR could be more clear:
In short, yes, "create multiple Send operators for each variable".
From the graph's perspective, the `Send` and `Recv` OP are one for each variable (but not one per replica, different (trainer) replicas share one `Send` and `Recv` for each variable).
In implementation, we could group send implementations to a single port handler.

In this design the variable placement is done by the *graph converter* before the graph is sent to worker, so it's not a runtime concept like a `Hash` operator. I think the "Hash" solution is for the simplest element-wise optimization case. If we want the parameter server to do things more than element-wise operation, we need to decide the parameter variable placement before the graph is sent to worker.

Reply:
Understood. You mean people who develop graphs, do not need to look into the implement details about how we actually send and recv variables, the graph is how the calculations flow logically. But when we build and optimize the graph, we can make the actual send operation one per trainer.

Will you add some implementation thoughts in this PR or in another one?

Reply:
@typhoonzero Thanks for the reminder on the implementation thoughts! That's a good idea. I will perhaps not mention implementation detail in this PR, but create a separate issue discussing it.